File: common/src/main/java/com/mthwate/conk/PositionUtils.java
package com.mthwate.conk;

import com.mthwate.datlib.math.vector.Vector3i;

/**
 * @author mthwate
 */
public class PositionUtils {

	public static int getLocalFromGlobal(int global, int chunkSize) {
		int n = global;

		if (n < 0) {
			n -= chunkSize - 1;
		}

		n %= chunkSize;

		if (global < 0) {
			n = chunkSize + n - 1;
		}

		return n;
	}

	public static Vector3i getLocalFromGlobal(int x, int y, int z, int chunkSize) {
		int lx = getLocalFromGlobal(x, chunkSize);
		int ly = getLocalFromGlobal(y, chunkSize);
		int lz = getLocalFromGlobal(z, chunkSize);
		return new Vector3i(lx, ly, lz);
	}

	public static Vector3i getLocalFromGlobal(Vector3i global, int chunkSize) {
		return getLocalFromGlobal(global.getX(), global.getY(), global.getZ(), chunkSize);
	}

	public static int getChunkFromGlobal(int global, int chunkSize) {
		if (global < 0) {
			global -= chunkSize - 1;
		}

		global /= chunkSize;

		return global;
	}

	public static Vector3i getChunkFromGlobal(int x, int y, int z, int chunkSize) {
		int cx = getChunkFromGlobal(x, chunkSize);
		int cy = getChunkFromGlobal(y, chunkSize);
		int cz = getChunkFromGlobal(z, chunkSize);
		return new Vector3i(cx, cy, cz);
	}

	public static Vector3i getChunkFromGlobal(Vector3i global, int chunkSize) {
		return getChunkFromGlobal(global.getX(), global.getY(), global.getZ(), chunkSize);
	}
}

File: common/src/main/kotlin/com/mthwate/conk/PositionUtils.kt
package com.mthwate.conk

import com.mthwate.datlib.math.vector.Vector3i

/**
 * @author mthwate
 */
object PositionUtils {

	fun getLocalFromGlobal(global: Int, chunkSize: Int): Int {
		var n = global

		if (n < 0) {
			n -= chunkSize - 1
		}

		n %= chunkSize

		if (global < 0) {
			n = chunkSize + n - 1
		}

		return n
	}

	fun getLocalFromGlobal(x: Int, y: Int, z: Int, chunkSize: Int): Vector3i {
		val lx = getLocalFromGlobal(x, chunkSize)
		val ly = getLocalFromGlobal(y, chunkSize)
		val lz = getLocalFromGlobal(z, chunkSize)
		return Vector3i(lx, ly, lz)
	}

	fun getLocalFromGlobal(global: Vector3i, chunkSize: Int): Vector3i {
		return getLocalFromGlobal(global.x, global.y, global.z, chunkSize)
	}

	fun getChunkFromGlobal(global: Int, chunkSize: Int): Int {
		var n = global
		if (n < 0) {
			n -= chunkSize - 1
		}

		n /= chunkSize

		return n
	}

	fun getChunkFromGlobal(x: Int, y: Int, z: Int, chunkSize: Int): Vector3i {
		val cx = getChunkFromGlobal(x, chunkSize)
		val cy = getChunkFromGlobal(y, chunkSize)
		val cz = getChunkFromGlobal(z, chunkSize)
		return Vector3i(cx, cy, cz)
	}

	fun getChunkFromGlobal(global: Vector3i, chunkSize: Int): Vector3i {
		return getChunkFromGlobal(global.x, global.y, global.z, chunkSize)
	}

}
